src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ComClassGenerator.cs

```diff
@@ -0,0 +1,225 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using static Microsoft.CodeAnalysis.CSharp.SyntaxFactory;
+using Microsoft.CodeAnalysis.CSharp;
+
+namespace Microsoft.Interop
+{
+    [Generator]
+    public class ComClassGenerator : IIncrementalGenerator
+    {
+        private sealed record ComClassInfo(string ClassName, ContainingSyntaxContext ContainingSyntaxContext, ContainingSyntax ClassSyntax, SequenceEqualImmutableArray<string> ImplementedInterfacesNames);
+        public void Initialize(IncrementalGeneratorInitializationContext context)
+        {
+            // Get all types with the [GeneratedComClassAttribute] attribute.
+            var attributedClasses = context.SyntaxProvider
+                .ForAttributeWithMetadataName(
+                    TypeNames.GeneratedComClassAttribute,
+                    static (node, ct) => node is ClassDeclarationSyntax,
+                    static (context, ct) =>
+                    {
+                        var type = (INamedTypeSymbol)context.TargetSymbol;
+                        var syntax = (ClassDeclarationSyntax)context.TargetNode;
+                        ImmutableArray<string>.Builder names = ImmutableArray.CreateBuilder<string>();
+                        foreach (INamedTypeSymbol iface in type.AllInterfaces)
+                        {
+                            if (iface.GetAttributes().Any(attr => attr.AttributeClass?.ToDisplayString() == TypeNames.GeneratedComInterfaceAttribute))
+                            {
+                                names.Add(iface.ToDisplayString());
+                            }
+                        }
+                        return new ComClassInfo(
+                            type.ToDisplayString(),
+                            new ContainingSyntaxContext(syntax),
+                            new ContainingSyntax(syntax.Modifiers, syntax.Kind(), syntax.Identifier, syntax.TypeParameterList),
+                            new(names.ToImmutable()));
+                    });
+
+            var classInfoType = attributedClasses.Select(static (info, ct) => GenerateClassInfoType(info)).SelectNormalized();
+
+            var attribute = attributedClasses.Select(static (info, ct) => GenerateClassInfoAttributeOnUserType(info)).SelectNormalized();
```

Comment:
I don't know what the `SelectNormalized()` does so this might not be helpful. Can we combine this into a single statement and return it as a tuple? It seems like we are going over the same collection twice.